### ISSUE TITLE
Support decoding and simulating AVX512 instructions

### ISSUE THREAD

**Issue by ltfish** [angr/angr#1386]
*Disclaimer: The angr suite is maintained by a small team of volunteers. While we cannot guarantee any timeliness for fixes and enhancements, we will do our best. For more real-time help with angr, from us and the community, join our [Slack.](http://angr.io/invite/)*

---

**Is your feature request related to a problem? Please describe.**

LibVEX does not support AVX512 instructions. As a result, angr does not support them, either.

**Describe the solution you'd like.**

Maybe applying the following patch?

- https://sourceforge.net/p/valgrind/mailman/message/36277875/
- https://bugs.kde.org/show_bug.cgi?id=383010

**Please include a sample of what *should* work if this feature is implemented.**

We will then be able to generate CFG on latest AMD64 libc.so.6 libraries.

**Describe alternatives you've considered.**

Don't use angr?

**Additional context**

Nope.

**Comment by zwimer** [comment]
@rhelmot If you are interested in this, assign it to yourself.

**Comment by zardus** [comment]
THIS MESSAGE WAS GENERATED BY AN AUTOMATED PROCESS

A census over a large binary corpus put numbers on this, and the population it
affects has changed shape: with Go's Green Tea GC this is no longer a libc
corner case.

### A reproducer that is already in the tree

`angr/binaries` at `a87538bc248531d2a1d22434143e8b9752459bac` already carries
binaries that hit this. `tests/x86_64/langdetect_gcc` is the easiest to check —
it statically links glibc's EVEX string routines:

```python
import angr
p = angr.Project("binaries/tests/x86_64/langdetect_gcc", auto_load_libs=False)
for n in ["__memset_avx512_unaligned", "__memset_avx512_unaligned_erms",
          "__memset_evex_unaligned", "__memset_evex_unaligned_erms",
          "__rawmemchr_evex", "__rawmemchr_evex_rtm",
          "__strchr_evex", "__strchrnul_evex"]:
    s = p.loader.find_symbol(n)
    b = p.factory.block(s.rebased_addr)
    print(f"{n:32s} {s.rebased_addr:#x} size={b.size} insns={b.instructions} {b.vex.jumpkind}")
```

```
__memset_avx512_unaligned        0x4237e0 size=0 insns=0 Ijk_NoDecode
__memset_avx512_unaligned_erms   0x423840 size=0 insns=0 Ijk_NoDecode
__memset_evex_unaligned          0x423a20 size=0 insns=0 Ijk_NoDecode
__memset_evex_unaligned_erms     0x423a80 size=0 insns=0 Ijk_NoDecode
__rawmemchr_evex                 0x424040 size=0 insns=0 Ijk_NoDecode
__rawmemchr_evex_rtm             0x4241c0 size=0 insns=0 Ijk_NoDecode
__strchr_evex                    0x42b1c0 size=0 insns=0 Ijk_NoDecode
__strchrnul_evex                 0x42bd40 size=0 insns=0 Ijk_NoDecode
```

The refused byte is the function's own first one: `62 e2 7d 48 7a c6` is
`vpbroadcastb zmm16, esi` and `62 e2 7d 28 7a c6` is `vpbroadcastb ymm16, esi`
(objdump 2.x and llvm-mc 21.1.8 agree). `CFGFast(normalize=True)` recovers 6,508
functions on that binary and **none of those eight** — the symbol table names
them, the bytes are ordinary code, and there is simply no function there.

The five sibling fixtures carry the same eight, at the same addresses:
`langdetect_gcc_dwarf`, `langdetect_clang`, `langdetect_rust` and
`langdetect_swift` through the symbol table, and `langdetect_gcc_stripped` —
which has no function symbols at all — through its 1,039 `.eh_frame` FDEs.

### Green Tea GC put AVX-512 in every amd64 Go binary

Go's Green Tea GC ships 26 hand-written AVX-512 routines,
`internal/runtime/gc/scan.expandAVX512_1` through `_64`, in
`src/internal/runtime/gc/scan/expand_amd64.s`. They are assembled
unconditionally for `GOARCH=amd64` and selected at run time, so they are linked
into the image **regardless of `GOAMD64`** — a default `GOAMD64=v1` build on a
machine that cannot execute AVX-512 still contains them. `GreenTeaGC: true` is in
the go1.27 baseline in `internal/buildcfg/exp.go`, and `GOEXPERIMENT=nogreenteagc`
is the only knob that removes them; the census also finds the same 26 in go1.26.4
and go1.26.5 builds.

A whole public reproducer, byte-reproducible with the stock toolchain:

```
$ cat go.mod
module avx512fixture

go 1.27
$ cat main.go
package main

func main() {}
$ CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -trimpath -buildvcs=false -ldflags="-s -w" -o fixture .
$ sha256sum fixture     # go1.27.0 linux/amd64
0fb5d60ddfbc8c4368059fbf35943e4712c8086a8df4ccfb4dda9f87ba8254ea  fixture   (1216672 bytes)
```

Its pclntab (parsed with `struct`, not through cle) declares 1,480 functions.
26 of them lift to `size=0 insns=0 Ijk_NoDecode`, and every one of the 26 is an
`expandAVX512_N`; no other declared start in the binary fails. `CFGFast`
recovers 1,743 functions, 0 of the 26, and no recovered function body covers any
of those 26 addresses. Stripping keeps all of it — Go retains the pclntab, so
the names survive with nothing behind them.

### Scope

Measured over a non-public corpus of 651,697 objects (counts and shapes only),
checking 122,930,403 function starts declared by sources parsed independently of
cle — ELF/PE/Mach-O symbol tables, Go pclntab, `.eh_frame` FDEs, PE `.pdata`, PE
exports, Mach-O `LC_FUNCTION_STARTS`, and DWARF:

* **4,824 objects / 125,060 declared function starts** refuse at the start
  because of EVEX. 3,194 objects and 82,826 starts were re-loaded and re-lifted
  in angr to confirm in context: 82,826 of 82,826, no disagreement. The
  unconfirmed remainder are Mach-O objects cle declines to load
  (`_load_lc_unixthread`, cle#727), not counter-evidence.
* **4,745 of the 4,824 are Go >= 1.26 amd64 builds** — 26 starts each. In the Go
  half of that corpus it is *every* amd64 Go binary: 4,672 of 4,672, across
  linux, darwin and windows targets, cross-checked by a raw byte search that
  does no ELF parsing and no disassembly at all.
* The other 79 are C, C++, Rust, Swift and Haskell binaries carrying EVEX
  library code: glibc's static EVEX string routines (8 starts each), Intel
  oneAPI's `__regcall3__svrng_coreavx512_*` and `_icv_k0_*` (65 and 32 starts
  each), and rustc/gcc/clang/ghc output.

One encoding, `62 f1 fe 48 6f 08` = `vmovdqu64 zmm1, [rax]`, accounts for
124,052 of the 125,060 sites.

### It is the whole EVEX space, not a feature subset

`guest_amd64_toIR.c` parses `0xC4` and `0xC5` only; there is no `0x62` branch,
and `grep -in evex` over it returns only `haveVEX`/`getVexL`, which are about
the *VEX* prefix. `VEX_HWCAPS_AMD64_AVX2 (1<<11)` is the highest x86 vector
capability in `libvex.h`. So `0x62` falls through to the legacy one-byte table
where it is invalid in 64-bit mode.

That means EVEX.128 and EVEX.256 are refused exactly as EVEX.512 is — the
census sees `vmovss xmm18`, `vpbroadcastb ymm16`, `vxorpd xmm16` and `vpxord
ymm16` among the refusals. Across the corpus, 6,946 distinct EVEX encodings were
refused, spanning AVX512F, AVX512BW, AVX512DQ, AVX512_VBMI, AVX512VL and GFNI.

Worth flagging for anyone scoping the work: **an EVEX-only decoder would not
close it.** The AVX-512 opmask instructions are VEX-encoded, and libVEX refuses
those too — `c4 e1 fb 92 c8`, `kmovq k1, rax`, is three-byte VEX and comes back
`Ijk_NoDecode`. Go's expanders use 22 of them.

### What it costs above the lifter

More than the missing functions, and none of it is visible at default log level:

* A neighbour that *is* recovered gets truncated. On the fixture above,
  `internal/runtime/gc/scan.scanSpanPackedAVX512` at `0x410280` is 288 bytes in
  the pclntab; `CFGFast` recovers 48 and stops at `0x4102b0`, where the bytes
  are `62 f1 fe 48 6f 18` — another `vmovdqu64`.
* Nothing marks the gap. The 26 expanders are reached through a pointer table;
  `kb.xrefs.get_xrefs_by_dst()` returns nothing for any of the 26, and no
  `kb.indirect_jumps` entry is recorded in the region, so there is no unresolved
  edge to follow either.
* Nothing survives the analysis. The segment list does record the range as
  `nodecode`, but it lives on the CFG analysis object, has no public accessor
  and is not serialized with `CFGModel`. What is left in the knowledge base is 26
  `kb.labels` entries with 0 functions behind them, and an `l.debug` line.
* The guard added in #6866 cannot reach this. It skips a function whose address
  the symbol table names, on exactly this reasoning — *"an instruction set the
  lifter does not implement looks exactly like data from here"* — but it iterates
  functions that already exist, and when the **first** instruction is the one
  that fails, `_generate_cfgnode` returns `None` and no function is ever created.
  A bodyless one would not survive either: `make_functions` deletes every
  function whose `startpoint is None`.

I am not proposing a CFG-side workaround for that; the honest fix is the one
this issue is already about.

session: sharpen

**Comment by zardus** [comment]
THIS MESSAGE WAS GENERATED BY AN AUTOMATED PROCESS

Two updates to my earlier comment.

**There is a tracked public reproducer now.** That comment gave a `go build` recipe because no committed binary showed the Go side of this. `tests/x86_64/windows/langdetect_go.exe` was added to `angr/binaries` on 2026-09-02 in commit b51d92b4, and it shows it directly, so nothing needs building:

```python
import angr
p = angr.Project("tests/x86_64/windows/langdetect_go.exe", auto_load_libs=False)
for s in p.loader.main_object.symbols:
    if s.name and "expandAVX512" in s.name:
        b = p.factory.block(s.rebased_addr, num_inst=1)
        print(s.name, hex(s.rebased_addr), b.size, b.vex.jumpkind)
```

All 26 `internal/runtime/gc/scan.expandAVX512_*` symbols come back `size=0` with `Ijk_NoDecode`. The first instruction is `62 f1 fe 48 6f ...`, an EVEX `vmovdqu64`. The eight symbols I listed in `tests/x86_64/langdetect_gcc` still behave the same way.

**The patch this issue was waiting on is not coming.** angr/pyvex#208 records the plan as merging "an official patch that at least make LibVEX decode AVX512 instructions", at very low priority. That is KDE bug 383010, filed in 2017 and still CONFIRMED with no resolution; the contributor stopped work on it. Checked today, neither angr/vex master (875f7c9) nor upstream Valgrind master has a `0x62` branch in the amd64 prefix parser, and `libvex.h` still tops out at `VEX_HWCAPS_AMD64_AVX2`.

I am not proposing a decoder. That is a large piece of work and a partial one would be worse than the clean refusal there is now. This is only to say that the reason recorded for waiting has expired.

**Comment by ltfish** [comment]
https://github.com/angr/angr/pull/7120 merges AVX512 support.
